### DISCUSSION TITLE
Updating a deeply nested state property: best practice

### DISCUSSION THREAD

**Discussion by kingjordan** [state-adapt/state-adapt#86] [Q&A]
Hello,

I am curious how others would approach this problem.  Say you have a property of your store that is a complex object (represented to the user as a dynamic form):

```javascript
interface myObjectState {
  sections: [
    index: number,
    {
      index: number;
      name: string;
      subSections: [
        {
          index: number;
          name: string;
          subSectionsRows: [
            {
              index: number;
              name: string;
              cells: [{ name: string; value: string }];
            }
          ];
        }
      ];
    }
  ];
}
```

example adapter:
```typescript
interface AppState {
  myObject: MyObjectType | null;
}

// set intital state
export const initialState: AppState = {
  myObject: null;
};

export const storeAdapter = createAdapter<AppState>()({
  ///...reducers
  setMyObject: (state, myOb) => ({ ...state, myObject: myOb }),
  selectors: {
    myObject: (s) => s.myObject,
  },
});
```

It can be quite a complex bit of code to update the value of a "cell" or add a row to a subSection inside of myObject.  I usually use immer's produce to accomplish this task.  I am wondering if state adapt has something built in for this task.

Secondly, would it be better to have a store that "normalizes" the deep object into top level properties of sections, subsections, rows and use id's to compose the object back together when needed?  

so the state for the object would look like:
```typescript
interface myObjectState {
  sections: [{ value: string }];
  subSections: [{ sectionId: number; value: string }];
  subSectionsRows: [
    { subSectionId: number; cells: [{ name: string; value: string }] }
  ];
}
```
and I would write a transform function/reducer to return the data in the nested structure:

```typescript
interface AppState {
  sections: [{ value: string }] | [];
  subSections: [{ sectionId: number; value: string }] | [];
  subSectionsRows: [
    { subSectionId: number; cells: [{ name: string; value: string }] }
  ] | [];
}

// set intital state
export const initialState: AppState = {
  sections: [],
  subSections: [],
  subSectionsRows: [],
};

export const storeAdapter = createAdapter<AppState>()({
  ///...reducers
  addSection: (state, section) => ({ ...state, sections: state.sections.push(section) }),
  selectors: {
    myObject: (s) => transFormMyObject(s), // use top level properties to compose myObject nested object
  },
});
```

Just looking for opinions on best practices in this scenario.

Thanks

**Comment by mfp22** [comment]
Flattening the state structure might be good, but if you have multiple tables on a page, I wouldn't combine their states. I used to think Redux could be like a client-side database, but it's usually best to just fetch fresh data from an endpoint rather than worrying about precise updates in a normalized client state. Offline mode might be different. But this is somewhat beside the point of your question. If it's just convenience, you have to weight the pros and cons, such as losing the ability to expand state the way it probably expands in the UI, and instead having to match up IDs. 

The usual answer for nested state updates is [`joinAdapters`](https://state-adapt.github.io/docs/core#joinadapters). Usually it's not just one-off updates you have to do, so having all the setup code defining the state structure is worth it. In your case, it's immediately worth it because it's less code:

```typescript
interface AppState {
  myObject: MyObjectType | null;
}

export const initialState: AppState = {
  myObject: null;
};

export const adapter = joinAdapters<AppState>()({
  myObject: createAdapter<MyObjectType | null>()({}),
})();
```

This will create `adapter.setMyObject` and `adapter.selectors.myObject` automatically. 

You could also create the inner adapter above to keep all the code organized by type. 

## Deeply Nested Arrays

At this moment I have no idea how to handle this. Normally `joinAdapters` is great. But with arrays, and _this many_ nested arrays, the state change functions that would be automatically generated might be pretty awkward. 

I need to come back to this. This might be the first scenario I've seen that turns out to really benefit from Immer over joinAdapters. But I need to think about it. In the meantime, take a look at `joinAdapters` and see if you can figure out a convenient way to use it for this.

One other pattern that might help is to create an array adapter creator:

```typescript
export function createArrayAdapter<T>() {
  return createAdapter<T[]>()({
    // ... reducers for arrays
  });
}
```

This is something StateAdapt should have soon. There's just been the entity adapter, but I imagine the issues with that would be worse with this. See this discussion: https://github.com/state-adapt/state-adapt/discussions/81

But even with this many nested arrays, that might be awkward too. The state change function might be `setSectionsSubSectionsRowsCellsOneValue`. For states this deeply nested, I almost wish I had gone with the alternate syntax that people on Twitter voted down: `sections_subSections_rows_cells_one_setValue`. Of course you can just rename that automatically generated reducer:

```typescript
export const adapter = joinAdapters<State>()({
  sections: sectionsAdapter,
})({
  setCellValue: sectionsAdapter.setSubSectionsRowsCellsOneValue,
})();
```

(Again see `joinAdapters` docs)

But if you're not doing much else, this is a lot of work for just one state change function. 

But the other issue would be the payload. The array adapter will automatically handle finding by id, so that needs to be separate from the payload. In the entity adapter, the overall action payload is like `[id, innerPayload]`. So you would end up with a payload like `[sectionId, [subSectionId, [rowId, [cellId, newValue]]]]`. That's pretty crazy! 

What are your thoughts on how Immer could be incorporated into StateAdapt? I'd prefer `joinAdapters` be more popular, but cases like this apparently exist.

**Reply by kingjordan** [reply]
Thanks for the follow up! I will try using joinAdapters and will give feedback on how that works.  

Secondly, the problem with immer is it's produce function requires a plain javascript object and the data in the rxjs stream is an observable.  So If I try to pass an action.paylod to produce it returns an error that it cannot accept anything other than primatives, arrays and plain objects.  So even if I wanted to use it I can't until I solve this issue:  [see this Stack Overflow question I created](https://stackoverflow.com/questions/78036359/immer-produce-inside-rxjs-stream-cant-operate-on-object?noredirect=1#comment137574361_78036359)  The type of action.payload is ```Observable {_subscribe: ƒ}``` and not ```{...properties}```. 

I am not sure if this is relevant, but the data I'm using here is a firestore doc with live updates, so I'm not sure how immer will affect that.  Currently I am testing most of this inside cypress component testing, I will connect it to real firestore data to see if that changes anything.

I agree that using the built in joinAdapters with state adapt feels like the better path so I will try that.  

Also, I read the linked discussion on arrayAdapters and I think it would be a useful feature.  I am not experienced enough to give great feedback on how it should work but I do agree that the id should probably be the array index.  For my scenario, the user would be re ordering the array sections/subsections/rows and adding + removing sections/subsections/rows is at the heart of the app's features.

**Reply by kingjordan** [reply]
i have set up a stack blitz to work through this issue here:
https://stackblitz.com/~/github.com/kingjordan/stackblitz-angular-state-adapt-demo

linked to this repo:
https://github.com/kingjordan/stackblitz-angular-state-adapt-demo

I have a basic working version of the nested data and components working, now I am adding state adapt to it

**Reply by kingjordan** [reply]
So i was able to set up an adapter via joinAdapters for the first level of nested structure:

```javascript
const sectionArrayAdapter = createAdapter<section[]>()({
    incrementIndexOfItem: (state, targetIndex: number)=> (moveElement(state, targetIndex, targetIndex + 1)),
    decrementIndexOfItem: (state, targetIndex: number)=> (moveElement(state, targetIndex, targetIndex - 1)),
    removeElementAtIndex: (state, targetIndex: number)=> (state.splice(targetIndex, 1)),
    addElementToEnd: (state, element: section)=> ([...state, element]),
    clearElements: () => ([]),
    selectors: {
        sectionLength: (s) => s.length,
        sectionElements: (s) => s
    }
})

export const appAdapter = joinAdapters<appState>()({
    title: baseStringAdapter,
    sections: sectionArrayAdapter
})();
```
and you can see it on[ this file of the repo:](https://github.com/kingjordan/stackblitz-angular-state-adapt-demo/blob/state-adapt/src/%2Bstate/app.adapter.ts).  checkout the state-adapt branch

I then create an app service to consume the adapter and create the top level app store:
```javascript
@Injectable({
    providedIn: "root",
  })
  export class AppService {

    private initialState: appState = {
        title: 'Initial State',
        sections: [section1, section2, section3]
    }

    private appStore = adapt(this.initialState, {
        path: "app",
        adapter: appAdapter
      });

    public appState$ = this.appStore.state$;

    constructor() {
        this.appStore.state$.pipe(takeUntilDestroyed()).subscribe();
    }
  }
```

and here is how I expose it to my app.component:
```javascript
@Component({
  standalone: true,
  imports: [CommonModule, SectionComponent],
  selector: 'demo-app',
  template: `
  <h1>Angular State Adapt Demo</h1>
  @if (appService.appState$ | async; as appState) {
    <h2>{{appState.title}}</h2>
    @if (appState.sections && appState.sections.length > 0){
      @for (section of appState.sections; track section.index; let idx = $index) {
        <div class="section-wrap">
          <demo-section [section]="section" />
        </div>
      }
    }
  }
  `,
})
export class AppComponent {
  public appService = inject(AppService);
}
```

it is nicely declarative.  Now I will try to add buttons so I can trigger adding/removing/moving a section via the adapters sources.

I am not sure how to use join adapters to represent the deeper nested data, in theory there should be an adapter<segment[]> that is nested inside of each adapter<section[]>.

**Reply by kingjordan** [reply]
Hi Mike,

I finished a working sample app trying to manage the nested sate.  [See github repo](https://github.com/kingjordan/stackblitz-angular-state-adapt-demo/tree/state-adapt)

[or open directly in stack blitz](https://stackblitz.com/~/github.com/kingjordan/stackblitz-angular-state-adapt-demo/tree/state-adapt)

I would be curious to get your feedback.  I tried to keep all logic in the parent app component and just bubble up the needed indexes through output events.  Would you think it makes sense to inject the app service in the lowest component to trigger actions or is it right to bubble up the needed info so that it can be handled by the parent/smart component?

The naming conventions were not bad, and I used a service to expose simpler action names to the other components.

**Reply by mfp22** [reply]
I think it makes sense to bubble up the actions, but it's hard to say.

I looked tat the adapter and it looks like you're using joinAdapters at the top level, but then you're doing mutations for deeply nested state after that. Is that the most recent code?

**Reply by kingjordan** [reply]
yeah, i could not figure out how to join adapters to represent the deeper state.  If i have a ```createAdapter<section[]>``` how do I join that to the ```segment[]``` array that exists on each section?  I don't need an adapter for each section which is what I think it would take to have access to the next level segments property of each section.  What do you think?

Its the array type that is causing the block for me.  I can easily join adapters on sub types if that type is represented by a single entity.

**Reply by kingjordan** [reply]
i tried to use the entity adapter to solve this problem:

```javascript
const rowsAdapter = createAdapter<row[]>()({});

const segmentAdapter = joinAdapters<segment, 'index'>()({
  title: baseStringAdapter,
  isCollapsed: booleanAdapter,
  rows: rowsAdapter,
});

const segmentEntityAdapter = createEntityAdapter<segment>()({});

const sectionAdapter = joinAdapters<section, 'index'>()({
  title: baseStringAdapter,
  isCollapsed: booleanAdapter,
  segments: segmentEntityAdapter,
});
```
But it throws an error on the ``` segments: segmentEntityAdapter,``` assignment:

> Type '{ addOne: (state: EntityState<segment, never>, payload: segment) => EntityState<segment, never>; addMany: (state: EntityState<segment, never>, payload: segment[]) => EntityState<...>; ... 7 more ...; upsertMany: (state: EntityState<...>, payload: segment[]) => EntityState<...>; } & ... 4 more ... & { ...; }' is not assignable to type '{ [index: string]: ((state: segment[], event: any, initialState: segment[]) => segment[]) | { [index: string]: (state: segment[]) => any; }; }'.
  Property 'addOne' is incompatible with index signature.
    Type '(state: EntityState<segment, never>, payload: segment) => EntityState<segment, never>' is not assignable to type '((state: segment[], event: any, initialState: segment[]) => segment[]) | { [index: string]: (state: segment[]) => any; }'.
      Type '(state: EntityState<segment, never>, payload: segment) => EntityState<segment, never>' is not assignable to type '(state: segment[], event: any, initialState: segment[]) => segment[]'.
        Types of parameters 'state' and 'state' are incompatible.
          Type 'segment[]' is missing the following properties from type 'EntityState<segment, never>': ids, entities

The problem is that I cannot assign the entity adapter to the array<segment> and that is the break point, we need a way to connect an adpater of array<T> or entity adpater to the collection property of the object.  
I dug through the createEntityAdadpter code to see if there was a return type I could use to type the segments property but as far as I can see there is no exposed entity type from the /core lib.

**Reply by mfp22** [reply]
The entity adapter documentation could use a lot of improvement, but given how complex your state already is, it's probably not the best fit. So the part under "Deeply Nested Arrays" in [this comment](https://github.com/state-adapt/state-adapt/discussions/86#discussioncomment-8553089) still applies. 

I wasn't sure what your paragraph about Immer was really about, but I just looked and saw that your Stack Overflow question was answered. Could you just use Immer for this?

**Reply by kingjordan** [reply]
Hi mike,

The question was answered in the way that I posted it.  However, immer cannot be used in this situation, at least I have not been able to yet.  The main reason is that the object passing through the state stream is an object of Action<T> and not Observable<T>.  So I am not able to "unwrap" the observable and pass a plain js object to immer which it requires.  Immer sees the action payload as a complex object and not a simple object.  Is there a way you can think of to "unwrap" the action payload into a plain object inside a rxjs stream?  That would allow immer to be used, however I was largely able to perform my state mutations inside the adapter even if it was a bit complex.  Immer would just clean up the code to be easier to read and maintain.  

On a side note,  bubbling up the events through 4-5 layers of inputs and outputs is becoming quite cumbersome.  I think I am leaning towards injecting the state service deeper into the nested components.  The only draw back I am finding is that I use cypress component testing for TDD and having to handle the state service is making that aspect more complicated.

**Reply by mfp22** [reply]
In a state adapter, you're receiving the action payload, not the action or an observable. And the payload is whatever you pass into the Source. Where are you trying to use Immer? I might be misunderstanding "Immer sees the action payload as a complex object" because the payload is coming from your code, not StateAdapt.
